geo/geomfn: implement ST_MaxDistance / ST_DFullyWithin

otan opened the pull request:
Based on top of #49085.

MaxDistance/DFullyWithin is a specialisation of distance, where it takes
the maximum of distances found instead and disregards intersections and
closest points. The `stopAfterLE` condition becomes `stopAfterGT`.

Resolves #48983, resolves #48916

Release note (sql change): Implemented the ST_MaxDistance and
ST_DFullyWithin function for geometries.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/49094)
<!-- Reviewable:end -->

sumeerbhola reviewed (changes requested):
> Reviewed 1 of 7 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49094)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @sumeerbhola)

___
*[pkg/geo/geomfn/distance.go, line 55 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49094#-M7PKGxC872hR-ahKhZm:-M7PKGxC872hR-ahKhZn:b-wdvknq) ([raw file](https://github.com/cockroachdb/cockroach/blob/b51da62b86bff3f96c59f2928b4b323ef25052db/pkg/geo/geomfn/distance.go#L55)):*
> ```Go
> 
> // DFullyWithin determines if any part of geometry A is fully within D units of
> // geometry B.
> ```

Is this just the maximum of the pair-wise min distance?
The PostGIS doc comment is quite useless "Returns true if the geometries is fully within the specified distance of one another."


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49094)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola)

___
*[pkg/geo/geomfn/distance.go, line 55 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49094#-M7PKGxC872hR-ahKhZm:-M7R1ok02slAvFz0FqzL:b-26ohdk) ([raw file](https://github.com/cockroachdb/cockroach/blob/b51da62b86bff3f96c59f2928b4b323ef25052db/pkg/geo/geomfn/distance.go#L55)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Is this just the maximum of the pair-wise min distance?
The PostGIS doc comment is quite useless "Returns true if the geometries is fully within the specified distance of one another."
</blockquote></details>

pair-wise distance. not quite "min distance", because intersection doesn't really count.
the better way to describe is "all points in geometry A is within D units of geometry B", but it can be thought of as MaxDistance(a,b) <= d.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49094)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/geo/geomfn/distance.go, line 55 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49094#-M7PKGxC872hR-ahKhZm:-M7c0c4O5QnD4f5Ylekd:bveu0gf) ([raw file](https://github.com/cockroachdb/cockroach/blob/b51da62b86bff3f96c59f2928b4b323ef25052db/pkg/geo/geomfn/distance.go#L55)):*
Can you add a code comment with a clear definition.

> pair-wise distance. not quite "min distance", because intersection doesn't really count.

what does "intersection doesn't really count" mean?

> the better way to describe is "all points in geometry A is within D units of geometry B", but it can be thought of as MaxDistance(a,b) <= d.

There is something about the geometric structure that allows almost the same algorithm to be used in that when doing max distance we still don't need to compare the interiors. Do you have thoughts on this and is it worth writing a code comment that justifies the algorithmic correctness?


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49094)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola)

___
*[pkg/geo/geomfn/distance.go, line 55 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49094#-M7PKGxC872hR-ahKhZm:-M7dg8IC0D2RgN5z8o13:b-agbhcb) ([raw file](https://github.com/cockroachdb/cockroach/blob/b51da62b86bff3f96c59f2928b4b323ef25052db/pkg/geo/geomfn/distance.go#L55)):*
> what does "intersection doesn't really count" mean?

never mind.

> There is something about the geometric structure that allows almost the same algorithm to be used in that when doing max distance we still don't need to compare the interiors. Do you have thoughts on this and is it worth writing a code comment that justifies the algorithmic correctness?

I've updated the comment to `maxDistanceInternal` since its algorithm related and applies to both these commands.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (changes requested):
> Reviewed 6 of 16 files at r1, 4 of 8 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49094)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @sumeerbhola)

___
*[pkg/geo/geodist/geodist.go, line 164 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49094#-M7h7tM1CzjTyzMezuKb:-M7h7tM1CzjTyzMezuKc:b-7tv918) ([raw file](https://github.com/cockroachdb/cockroach/blob/c41b28af239f4b6ab5408194b458bac31978de91/pkg/geo/geodist/geodist.go#L164)):*
> ```Go
> 			return true
> 		}
> 		if !c.DistanceUpdater().IsMaxDistance() {
> ```

I don't see why this is correct -- I thought we wanted the minimum distance even in this case, and later do the max of the min distance between shapes. If this is being used where the point is not a shape in the original slices then it may be a different calculation.
___
*[pkg/geo/geodist/geodist.go, line 192 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49094#-M7h72uf-F_PfhsICbxk:-M7h72uf-F_PfhsICbxl:b4kx1ia) ([raw file](https://github.com/cockroachdb/cockroach/blob/c41b28af239f4b6ab5408194b458bac31978de91/pkg/geo/geodist/geodist.go#L192)):*
> ```Go
> 	// If the exterior ring does not contain the point, we just need to calculate the distance to
> 	// the exterior ring.
> 	// Also, if we are just calculating max distance, we only want the distance to the exterior.
> ```

I don't understand this. If the point is contained in the polygon, isn't it incorrect to measure its distance from the exterior. The max distance is 0, yes?
Also, it may not be contained in the polygon, but contained in a hole, in which case the max distance will be the distance to the hole.
In short, the distance calculation where one shape is a point should be the same for max and min
___
*[pkg/geo/geodist/geodist.go, line 226 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49094#-M7hC3uT98mLW4OLJ_f1:-M7hC3uT98mLW4OLJ_f2:bmxs8vj) ([raw file](https://github.com/cockroachdb/cockroach/blob/c41b28af239f4b6ab5408194b458bac31978de91/pkg/geo/geodist/geodist.go#L226)):*
> ```Go
> 
> 			// Compare each vertex against the edge of the other.
> 			for _, toCheck := range []struct {
> ```

I don't see why this decomposition is correct when being called from `ShapeDistance()`.
I think we want
max over all i (min over all j (distance(vertex a[i], edge b[j])) 
___
*[pkg/geo/geodist/geodist.go, line 265 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49094#-M7h967Y9cN0o1Pl8XQb:-M7h967Y9cN0o1Pl8XQc:b-3gu5mi) ([raw file](https://github.com/cockroachdb/cockroach/blob/c41b28af239f4b6ab5408194b458bac31978de91/pkg/geo/geodist/geodist.go#L265)):*
> ```Go
> 	// We use the first point of the linestring for this check.
> 	// If we are looking for max distance, we only need to check against the outer ring anyway.
> 	if c.DistanceUpdater().IsMaxDistance() || !c.PointInLinearRing(a.Vertex(0), b.LinearRing(0)) {
> ```

same question as with the point and polygon comparison.
___
*[pkg/geo/geomfn/distance.go, line 55 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49094#-M7PKGxC872hR-ahKhZm:-M7hAdbi8MrC_9zyTgPz:b4x9c27) ([raw file](https://github.com/cockroachdb/cockroach/blob/b51da62b86bff3f96c59f2928b4b323ef25052db/pkg/geo/geomfn/distance.go#L55)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

> what does "intersection doesn't really count" mean?

never mind.

> There is something about the geometric structure that allows almost the same algorithm to be used in that when doing max distance we still don't need to compare the interiors. Do you have thoughts on this and is it worth writing a code comment that justifies the algorithmic correctness?

I've updated the comment to `maxDistanceInternal` since its algorithm related and applies to both these commands.
</blockquote></details>

I still don't see why this decomposition is correct:
If A is composed of shapes A1, A2 and B is composed of shapes B1, B2
Say every point in B1 is within 1 unit of A1 and every point in B2 is within 1 unit of A2.
But every point in B1 is within 10 units of A2 and every point in B2 is within 10 units of A1.
I am assuming the max distance between A and B is 1.
But if we calculate every pair-wise distance and take the max we will get 10.
Am I missing something?


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49094)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola)

___
*[pkg/geo/geodist/geodist.go, line 164 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49094#-M7h7tM1CzjTyzMezuKb:-M7hHShY8Xi3MerbhNJu:b-o77fqk) ([raw file](https://github.com/cockroachdb/cockroach/blob/c41b28af239f4b6ab5408194b458bac31978de91/pkg/geo/geodist/geodist.go#L164)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I don't see why this is correct -- I thought we wanted the minimum distance even in this case, and later do the max of the min distance between shapes. If this is being used where the point is not a shape in the original slices then it may be a different calculation.
</blockquote></details>

Ok, I think that comment is wrong / the understanding of max distance is different between us.

We don't actually want the minimum distance - not sure how to describe it with eloquent math, but basically you want to find the distance between every point of everything in A and everything in B, and take the maximum of that.

Let's take for example max distance of the same polygon in postgis:
```
otan=# select ST_MaxDistance('POLYGON((0.0 0.0, 1.0 0.0, 1.0 1.0, 0.0 1.0, 0.0 0.0))',  'POLYGON((0.0 0.0, 1.0 0.0, 1.0 1.0, 0.0 1.0, 0.0 0.0))');
   st_maxdistance
--------------------
 1.4142135623730951
(1 row)
```

the maximum distance is from (0.0 0.0) to (1.0 1.0), which is sqrt(2). even if things are the same polygon, to be "fully within" they must be within sqrt(2) with each other.
even in the case of multiple polygons, this calculation is still correct. note that it is not 0, that's the minimum distance.

perhaps the test cases will help give an idea of the calculations -- note that i generated those cases by plugging the output of PostGIS into the "expected".
___
*[pkg/geo/geodist/geodist.go, line 192 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49094#-M7h72uf-F_PfhsICbxk:-M7hILEp8gXUOYTd06bt:b-dgehmx) ([raw file](https://github.com/cockroachdb/cockroach/blob/c41b28af239f4b6ab5408194b458bac31978de91/pkg/geo/geodist/geodist.go#L192)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I don't understand this. If the point is contained in the polygon, isn't it incorrect to measure its distance from the exterior. The max distance is 0, yes?
Also, it may not be contained in the polygon, but contained in a hole, in which case the max distance will be the distance to the hole.
In short, the distance calculation where one shape is a point should be the same for max and min
</blockquote></details>

No, max distance is never 0 unless its just two points.
This is correct because the exterior rings are always guaranteed to be the things furthest away from each other.
___
*[pkg/geo/geodist/geodist.go, line 226 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49094#-M7hC3uT98mLW4OLJ_f1:-M7hIwwR-oqQd0U9AkSa:b-bexu4y) ([raw file](https://github.com/cockroachdb/cockroach/blob/c41b28af239f4b6ab5408194b458bac31978de91/pkg/geo/geodist/geodist.go#L226)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I don't see why this decomposition is correct when being called from `ShapeDistance()`.
I think we want
max over all i (min over all j (distance(vertex a[i], edge b[j])) 
</blockquote></details>

See first answer.
___
*[pkg/geo/geodist/geodist.go, line 265 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49094#-M7h967Y9cN0o1Pl8XQb:-M7hJB2G3BY7Z113RgeZ:b-em0zb3) ([raw file](https://github.com/cockroachdb/cockroach/blob/c41b28af239f4b6ab5408194b458bac31978de91/pkg/geo/geodist/geodist.go#L265)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

same question as with the point and polygon comparison.
</blockquote></details>

See answer above.
___
*[pkg/geo/geomfn/distance.go, line 55 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49094#-M7PKGxC872hR-ahKhZm:-M7hIU6HE4wUhmYryKaU:bsjte78) ([raw file](https://github.com/cockroachdb/cockroach/blob/b51da62b86bff3f96c59f2928b4b323ef25052db/pkg/geo/geomfn/distance.go#L55)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I still don't see why this decomposition is correct:
If A is composed of shapes A1, A2 and B is composed of shapes B1, B2
Say every point in B1 is within 1 unit of A1 and every point in B2 is within 1 unit of A2.
But every point in B1 is within 10 units of A2 and every point in B2 is within 10 units of A1.
I am assuming the max distance between A and B is 1.
But if we calculate every pair-wise distance and take the max we will get 10.
Am I missing something?
</blockquote></details>

See above comment.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (changes requested):
> Reviewed 1 of 16 files at r1, 1 of 7 files at r2, 1 of 8 files at r3, 4 of 4 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49094)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @sumeerbhola)

___
*[docs/generated/sql/functions.md, line 753 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/49094#-M7j0FUVEUm4v-DL909D:-M7j0FUVEUm4v-DL909E:b-lnl7ev) ([raw file](https://github.com/cockroachdb/cockroach/blob/083327743711dd456b918aeccd89e2b5e3350516/docs/generated/sql/functions.md#L753)):*
> ```Markdown
> <p>This function will automatically use any available index.</p>
> </span></td></tr>
> <tr><td><a name="st_dfullywithin"></a><code>st_dfullywithin(geometry_a: geometry, geometry_b: geometry, distance: <a href="float.html">float</a>) &rarr; <a href="bool.html">bool</a></code></td><td><span class="funcdesc"><p>Returns true if all of geometry_a is in distance units of geometry_b. In other words, the max distance between geometry_a and geometry_b is less than distance units.</p>
> ```

This is open to interpretation. How about:
Returns true if every pair of points comprising geometry_a and geometry_b respectively, are within distance units.
___
*[docs/generated/sql/functions.md, line 870 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/49094#-M7j1-_rD7Aco4ypV9x6:-M7j1-_rD7Aco4ypV9x7:bxpsivj) ([raw file](https://github.com/cockroachdb/cockroach/blob/083327743711dd456b918aeccd89e2b5e3350516/docs/generated/sql/functions.md#L870)):*
> ```Markdown
> <tr><td><a name="st_makepoint"></a><code>st_makepoint(x: <a href="float.html">float</a>, y: <a href="float.html">float</a>) &rarr; geometry</code></td><td><span class="funcdesc"><p>Returns a new Point with the given X and Y coordinates.</p>
> </span></td></tr>
> <tr><td><a name="st_maxdistance"></a><code>st_maxdistance(geometry_a: geometry, geometry_b: geometry) &rarr; <a href="float.html">float</a></code></td><td><span class="funcdesc"><p>Returns the maximum distance between the given geometries. Note if the geometries are the same, it will return the maximum distance between the geometry’s vertexes.</p>
> ```

Can we make this more specific. Something like:
Returns the maximum distance across every pair of points comprising the given geometries. Note if the ...
___
*[pkg/geo/geodist/geodist.go, line 192 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49094#-M7h72uf-F_PfhsICbxk:-M7j3wsP17TkY5mh0voV:b-g8u962) ([raw file](https://github.com/cockroachdb/cockroach/blob/c41b28af239f4b6ab5408194b458bac31978de91/pkg/geo/geodist/geodist.go#L192)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

No, max distance is never 0 unless its just two points.
This is correct because the exterior rings are always guaranteed to be the things furthest away from each other.
</blockquote></details>

This is somewhat nuanced and deserves a comment. I managed to convince myself based on breaking down the cases as follows, but if you can find a reference that would be better.
// When computing the maximum distance, the cases are:
// - The point P is not contained in the exterior of the polygon G. Say vertex V is the vertex of the exterior of the polygon that is furthest away from point P (among all the exterior vertices). One can prove that any vertex of the holes will be closer to point P than vertex V. Similarly we can prove that any point in the interior of the polygin is closer to P than vertex V. Therefore we only need to compare with the exterior.
// - The point P is contained in the exterior and inside a hole of polygon G. One can again prove that the furthest point in the polygon from P is one of the vertices of the exterior.
// - The point P is contained in the polygon. One can again prove the same property.
// So we only need to compare with the exterior.
___
*[pkg/geo/geodist/geodist.go, line 156 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/49094#-M7j3Q37EvGLg5eH5XIN:-M7j3Q37EvGLg5eH5XIO:b-a80wcx) ([raw file](https://github.com/cockroachdb/cockroach/blob/083327743711dd456b918aeccd89e2b5e3350516/pkg/geo/geodist/geodist.go#L156)):*
> ```Go
> 
> // onPointToEdgesExceptFirstEdgeStart updates the distance using the edges between a point and a shape.
> // It will only check the ends of the edges, and assumes the check against .Edge(0).V0 is not required.
> ```

The "It will only check the ends of the edges" doesn't reflect the changed function name from the previous PR.
___
*[pkg/geo/geodist/geodist.go, line 164 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/49094#-M7j394h0CcngUe_Ei9R:-M7j394h0CcngUe_Ei9S:b-y2w0o2) ([raw file](https://github.com/cockroachdb/cockroach/blob/083327743711dd456b918aeccd89e2b5e3350516/pkg/geo/geodist/geodist.go#L164)):*
> ```Go
> 			return true
> 		}
> 		if !c.DistanceUpdater().IsMaxDistance() {
> ```

Can you add a comment before the if-block
// Max distance between a point and the set of points representing an edge is the maximum distance from the point and the pair of end-points of the edge, so we don't need update the distance using the projected point.
___
*[pkg/geo/geodist/geodist.go, line 217 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/49094#-M7j7X7L3sR8MsasXi5T:-M7j7X7L3sR8MsasXi5U:b-7x1bj6) ([raw file](https://github.com/cockroachdb/cockroach/blob/083327743711dd456b918aeccd89e2b5e3350516/pkg/geo/geodist/geodist.go#L217)):*
> ```Go
> 		crosser := c.NewEdgeCrosser(aEdge, b.Edge(0).V0)
> 		for bEdgeIdx := 0; bEdgeIdx < b.NumEdges(); bEdgeIdx++ {
> 			bEdge := b.Edge(bEdgeIdx)
> ```

// Max distance between 2 edges is the maximum of the distance across pairs of vertices chosen from each edge.
// It does not matter whether the edges cross, so we skip that check.
___
*[pkg/geo/geodist/geodist.go, line 264 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/49094#-M7j83Ms1YgVNPmzBk2l:-M7j83Ms1YgVNPmzBk2m:b-myyp1o) ([raw file](https://github.com/cockroachdb/cockroach/blob/083327743711dd456b918aeccd89e2b5e3350516/pkg/geo/geodist/geodist.go#L264)):*
> ```Go
> 	// and the exterior ring.
> 	// We use the first point of the linestring for this check.
> 	// If we are looking for max distance, we only need to check against the outer ring anyway.
> ```

Let's update this comment as
// Min Distance:
// ....
// Max Distance:
// The max distance is the distance between the line string and the exterior ring of the polygon (using the same reasoning as described in onPointToPolygon()
___
*[pkg/geo/geodist/geodist.go, line 318 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/49094#-M7j8k_O8LT4nlRa16w_:-M7j8k_O8LT4nlRa16wa:b-bgg8ej) ([raw file](https://github.com/cockroachdb/cockroach/blob/083327743711dd456b918aeccd89e2b5e3350516/pkg/geo/geodist/geodist.go#L318)):*
> ```Go
> 	// As such, we only need to compare the exterior rings if we detect this.
> 	//
> 	// If we are only looking at the max distance, we only want to compare exteriors.
> ```

Like my comment on `onPointToPolygon` this is non-obvious. Do you happen to have a reference?
___
*[pkg/geo/geomfn/distance.go, line 55 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49094#-M7PKGxC872hR-ahKhZm:-M7j1n1y113TY6hFKACi:bdxr4y7) ([raw file](https://github.com/cockroachdb/cockroach/blob/b51da62b86bff3f96c59f2928b4b323ef25052db/pkg/geo/geomfn/distance.go#L55)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

See above comment.
</blockquote></details>

How about:

// DFullyWithin determines whether the maximum distance across every pair of points comprising geometries A and B is within D units.
___
*[pkg/geo/geomfn/distance.go, line 31 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/49094#-M7j1Yov-wiGNmiz4UDd:-M7j1Yow4JiUcf1OIfsI:b-79mu7n) ([raw file](https://github.com/cockroachdb/cockroach/blob/083327743711dd456b918aeccd89e2b5e3350516/pkg/geo/geomfn/distance.go#L31)):*
> ```Go
> }
> 
> // MaxDistance returns the maximum distance between geometries A and B.
> ```

How about:
// MaxDistance returns the maximum distance across every pair of points comprising geometries A and B.


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49094)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola)

___
*[docs/generated/sql/functions.md, line 753 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/49094#-M7j0FUVEUm4v-DL909D:-M7jMHR08huGFw0fq8q-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/083327743711dd456b918aeccd89e2b5e3350516/docs/generated/sql/functions.md#L753)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

This is open to interpretation. How about:
Returns true if every pair of points comprising geometry_a and geometry_b respectively, are within distance units.
</blockquote></details>

Done.
___
*[docs/generated/sql/functions.md, line 870 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/49094#-M7j1-_rD7Aco4ypV9x6:-M7jMH6S5So1Vq7zsqdg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/083327743711dd456b918aeccd89e2b5e3350516/docs/generated/sql/functions.md#L870)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Can we make this more specific. Something like:
Returns the maximum distance across every pair of points comprising the given geometries. Note if the ...
</blockquote></details>

Done.
___
*[pkg/geo/geodist/geodist.go, line 192 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49094#-M7h72uf-F_PfhsICbxk:-M7jNwDoAO06XX7ZrIdQ:bwz61vz) ([raw file](https://github.com/cockroachdb/cockroach/blob/c41b28af239f4b6ab5408194b458bac31978de91/pkg/geo/geodist/geodist.go#L192)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

This is somewhat nuanced and deserves a comment. I managed to convince myself based on breaking down the cases as follows, but if you can find a reference that would be better.
// When computing the maximum distance, the cases are:
// - The point P is not contained in the exterior of the polygon G. Say vertex V is the vertex of the exterior of the polygon that is furthest away from point P (among all the exterior vertices). One can prove that any vertex of the holes will be closer to point P than vertex V. Similarly we can prove that any point in the interior of the polygin is closer to P than vertex V. Therefore we only need to compare with the exterior.
// - The point P is contained in the exterior and inside a hole of polygon G. One can again prove that the furthest point in the polygon from P is one of the vertices of the exterior.
// - The point P is contained in the polygon. One can again prove the same property.
// So we only need to compare with the exterior.
</blockquote></details>

I don't think it's a reference and it seems somewhat intuitive to me (I've been starting at it for a long time HAH!) -- I'll add this comment.
___
*[pkg/geo/geodist/geodist.go, line 226 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49094#-M7hC3uT98mLW4OLJ_f1:-M7jNuC2C8H1lUHMlnvg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c41b28af239f4b6ab5408194b458bac31978de91/pkg/geo/geodist/geodist.go#L226)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

See first answer.
</blockquote></details>

Done.
___
*[pkg/geo/geodist/geodist.go, line 265 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49094#-M7h967Y9cN0o1Pl8XQb:-M7jS-lODPahK-8ADen8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c41b28af239f4b6ab5408194b458bac31978de91/pkg/geo/geodist/geodist.go#L265)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

See answer above.
</blockquote></details>

Done.
___
*[pkg/geo/geodist/geodist.go, line 156 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/49094#-M7j3Q37EvGLg5eH5XIN:-M7jOfTBC04aAx2GjmFM:bcx20mn) ([raw file](https://github.com/cockroachdb/cockroach/blob/083327743711dd456b918aeccd89e2b5e3350516/pkg/geo/geodist/geodist.go#L156)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

The "It will only check the ends of the edges" doesn't reflect the changed function name from the previous PR.
</blockquote></details>

How so? I've updated it slightly.
___
*[pkg/geo/geodist/geodist.go, line 164 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/49094#-M7j394h0CcngUe_Ei9R:-M7jSJwt7HzFSgATGTHl:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/083327743711dd456b918aeccd89e2b5e3350516/pkg/geo/geodist/geodist.go#L164)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Can you add a comment before the if-block
// Max distance between a point and the set of points representing an edge is the maximum distance from the point and the pair of end-points of the edge, so we don't need update the distance using the projected point.
</blockquote></details>

Done.
___
*[pkg/geo/geodist/geodist.go, line 217 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/49094#-M7j7X7L3sR8MsasXi5T:-M7jRERj9rg2uOGCZqV2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/083327743711dd456b918aeccd89e2b5e3350516/pkg/geo/geodist/geodist.go#L217)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

// Max distance between 2 edges is the maximum of the distance across pairs of vertices chosen from each edge.
// It does not matter whether the edges cross, so we skip that check.
</blockquote></details>

Done.
___
*[pkg/geo/geodist/geodist.go, line 264 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/49094#-M7j83Ms1YgVNPmzBk2l:-M7jRzrMBQ-WU4dMsx7a:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/083327743711dd456b918aeccd89e2b5e3350516/pkg/geo/geodist/geodist.go#L264)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Let's update this comment as
// Min Distance:
// ....
// Max Distance:
// The max distance is the distance between the line string and the exterior ring of the polygon (using the same reasoning as described in onPointToPolygon()
</blockquote></details>

Done.
___
*[pkg/geo/geodist/geodist.go, line 318 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/49094#-M7j8k_O8LT4nlRa16w_:-M7jS03O-xfMpKMHNDFU:b-ic4ex8) ([raw file](https://github.com/cockroachdb/cockroach/blob/083327743711dd456b918aeccd89e2b5e3350516/pkg/geo/geodist/geodist.go#L318)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Like my comment on `onPointToPolygon` this is non-obvious. Do you happen to have a reference?
</blockquote></details>

Nope. But your analysis is correct - I've updated the comment
___
*[pkg/geo/geomfn/distance.go, line 55 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49094#-M7PKGxC872hR-ahKhZm:-M7jMVcE4RyTaAmJEEkj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b51da62b86bff3f96c59f2928b4b323ef25052db/pkg/geo/geomfn/distance.go#L55)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

How about:

// DFullyWithin determines whether the maximum distance across every pair of points comprising geometries A and B is within D units.

</blockquote></details>

Done.
___
*[pkg/geo/geomfn/distance.go, line 31 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/49094#-M7j1Yov-wiGNmiz4UDd:-M7jMVvOBdtjNXyekMy6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/083327743711dd456b918aeccd89e2b5e3350516/pkg/geo/geomfn/distance.go#L31)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

How about:
// MaxDistance returns the maximum distance across every pair of points comprising geometries A and B.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 4 of 4 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49094#-:-M7ouTo60wE59mCFFeH_:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @otan and @sumeerbhola)

___
*[pkg/geo/geodist/geodist.go, line 192 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49094#-M7h72uf-F_PfhsICbxk:-M7orkuZ08QYxiiCKrPG:bxh02kp) ([raw file](https://github.com/cockroachdb/cockroach/blob/c41b28af239f4b6ab5408194b458bac31978de91/pkg/geo/geodist/geodist.go#L192)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

I don't think it's a reference and it seems somewhat intuitive to me (I've been starting at it for a long time HAH!) -- I'll add this comment.
</blockquote></details>

I've found intuition (both mine and others) to be incorrect enough times. Given the complexity of holes and concave polygons, intuition can be tripped up here. But this is fine for now.
___
*[pkg/geo/geodist/geodist.go, line 167 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/49094#-M7orZA_18hEXP7AtteF:-M7orZA_18hEXP7AtteG:b-afx2lz) ([raw file](https://github.com/cockroachdb/cockroach/blob/7eb62385b5b101ff1b85fba4ba7972cdb7df8587/pkg/geo/geodist/geodist.go#L167)):*
> ```Go
> 		// The max distance between a point and the set of points representing an edge is the
> 		// maximum distance from the point and the pair of end-points of the edge, so we don't
> 		// need update the distance using the projected point.
> ```

... need to update ...


<!-- Sent from Reviewable.io -->

otan committed:
geo/geomfn: implement ST_MaxDistance / ST_DFullyWithin

MaxDistance/DFullyWithin is a specialisation of distance, where it takes
the maximum of distances found instead and disregards intersections and
closest points. The `stopAfterLE` condition becomes `stopAfterGT`.

Release note (sql change): Implemented the ST_MaxDistance and
ST_DFullyWithin function for geometries.

otan commented:
tftr!

bors r=sumeerbhola
